exploration/0477-quoted-literals.md

```diff
@@ -0,0 +1,268 @@
+
+- **[r3; medium priority]** Minimize the incentive to avoid escaping by changing messages
+  (e.g. rephrasing content, using typographic apostrophes, or switching outer delimiters).
+
+- **[r4; medium priority]** Don't surprise users with syntax that's too exotic.
+  We expect quoted literals to be rare,
+  which means fewer opportunities to get used to their syntax and remember it.
+
+- **[r5; low priority]** Be able to pair the opening and the closing delimiter,
+  to aid parsers recover from syntax errors,
+  and to leverage IDE's ability to highlight matching pairs of delimiters,
+  to visually indicate to the user editing a message the bounds of the literal under caret.
+  However, quoted literals are usually short and already enclosed in a placeholder (which has its own delimiters)
+  or are outside patterns (when used as variant keys).
+
+## Constraints
+
+_What prior decisions and existing conditions limit the possible design?_
+
+- **[c1]** MessageFormat uses the backslash, `\`,
+  as the escape sequence introducer.
+
+- **[c2]** Straight quotation marks, `'` and `"`,
+  are common in content across many languages,
+  even if other Unicode codepoints should be used in well-formatted text.
+
+- **[c3]** Straight quotation marks, `'` and `"`,
+  are common as string delimiters in many programming languages.
+
+## Proposed Design
+
+_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._
+
+Use the vertical line character, `|`, to delimit quoted strings. For example:
+
+> ```
+> {The Unix epoch is defined as {|Thu, 01 Jan 1970 00:00:00 GMT| :datetime}.}
+> ```
+
+```abnf
+literal       = quoted / unquoted
+quoted        = "|" *(quoted-char / quoted-escape) "|"
+quoted-char   = %x0-5B         ; omit \
+              / %x5D-7B        ; omit |
+              / %x7D-D7FF      ; omit surrogates
+              / %xE000-10FFFF
+quoted-escape = backslash ( backslash / "|" )
+```
+
+By being both uncommon in text content
+and uncommon as a string delimiter in other programming languages,
+the vertical line sidesteps the "inwards" and "outwards" problems of escaping.
+
+- [r1 GOOD] Writing `"` and `'` in literals doesn't require escaping them via `\`.
+  This means no extra `\` that need escaping.
+- [r2 GOOD] Embedding messages in most code or containers doesn't require escaping the literal delimiters.
+- [r3 GOOD] Message don't have to be modified otherwise before embedding them.
+- [r4 FAIR] Vertical lines are not commonly used as string delimiters
+  and thus can be harder to learn for beginners.
+  There's prior art in a practice of using vertical lines as delimiters for inline code literals in email and chat.
+  Vertical bars can also be used to delimit [regular expressions in `sed`](https://en.wikipedia.org/wiki/Vertical_bar#Delimiter),
+  and as a separator in [delimiter-separated data formats](http://www.catb.org/~esr/writings/taoup/html/ch05s02.html).
```

Comment:
For sed, this argument can be made about [almost any character](https://pubs.opengroup.org/onlinepubs/009604599/utilities/sed.html):
> Any character other than backslash or \<newline> can be used instead of a slash to delimit the BRE and the replacement.

I, for one, tend to use `#` when the regexp includes `/` characters.

This is not really valid prior art for `|`.

You may want to also consider markdown tables as a "delimiter" use of `|`, though I'd say that's about as much a stretch as its use as a DSV separator.

```suggestion
  Vertical bars can also be used as a separator in [delimiter-separated data formats](http://www.catb.org/~esr/writings/taoup/html/ch05s02.html).
```

Reply:
Your comment made me realize that vertical lines are indeed more often used as delimiters for _separating_ rather than for _enclosing_.